### PR TITLE
[EagerAppCDS] enable classpath change by replacing given env var

### PR DESCRIPTION
**Pull request by jia-wei-tang**
Summary: If user give environment vars and use
`-Dcom.alibaba.cds.cp.reloc.envs`, cds will change the classpath by using given environment vars to transform some paths in classpath.

Test Plan: ci jtreg
jdk/test/com/alibaba/quickstart/TestEnvVariableReplay.java jdk/test/com/alibaba/quickstart/TestEnvVariableDump.java

Reviewed-by: lingjun-cg, yuleil

Issue: https://github.com/dragonwell-project/dragonwell8/issues/595

### CODE REVIEW

File: hotspot/src/share/vm/runtime/quickStart.cpp [hunk 1]
```diff
@@ -1006,7 +1048,117 @@ bool QuickStart::dump_cached_info(const JavaVMInitArgs* options_args) {
...
+}
+
+char* QuickStart::replace_if_contains(const char* path, bool* replaced) {
+  if(old_envs != NULL) {
+    int array_cnt = old_envs->length();
+    for(int i = 0; i < array_cnt; i++) {
+      const char* target = old_envs->at(i);
+      if(strstr(path, target) != NULL) {
+        int target_len = strlen(target);
+        const char* new_target = new_envs->at(i);
```
Comment (lingjun-cg) [inline hotspot/src/share/vm/runtime/quickStart.cpp:1132]: 替换成StringUtils::replace_no_expand吧

File: hotspot/src/share/vm/runtime/quickStart.cpp [hunk 2]
```diff
@@ -1006,7 +1048,117 @@ bool QuickStart::dump_cached_info(const JavaVMInitArgs* options_args) {
...
+  bool process_end = false;
+  while (!process_end) {
+    if (origin_path[p2] == '\0') {
+      process_end = true;
+    }
+    if((origin_path[p2] == *os::path_separator() || origin_path[p2] == '\0')) {
+      char split[JVM_MAXPATHLEN];
+      strncpy(split, &origin_path[p1], p2-p1);
+      split[p2-p1] = '\0';
+      bool replaced = false;
```
Comment (lingjun-cg) [inline hotspot/src/share/vm/runtime/quickStart.cpp:1094]: replaced似乎没有用到？

File: hotspot/src/share/vm/runtime/quickStart.cpp [hunk 3]
```diff
@@ -1006,7 +1048,117 @@ bool QuickStart::dump_cached_info(const JavaVMInitArgs* options_args) {
...
+        strcat(new_path, new_target);
+        strcat(new_path, &path[target_len]);
+        strcat(new_path, "\0");
+        *replaced = true;
+        return new_path;
+      }
+    }
+  }
+
+  char* new_path = NEW_C_HEAP_ARRAY(char, strlen(path) + 1, mtInternal);
```
Comment (lingjun-cg) [inline hotspot/src/share/vm/runtime/quickStart.cpp:1145]: 如果不需要替换的话，这块不需要分配内存

**Review by lingjun-cg — COMMENTED** [review]
参考具体文件的comments

File: hotspot/src/share/vm/runtime/quickStart.cpp (hunk 2 above)
Comment (jia-wei-tang) [reply hotspot/src/share/vm/runtime/quickStart.cpp:1094]: 有两处需要这个函数，replace是为了filemap.cpp中的使用。

File: hotspot/src/share/vm/runtime/quickStart.cpp (hunk 1 above)
Comment (jia-wei-tang) [reply hotspot/src/share/vm/runtime/quickStart.cpp:1132]: replace_no_expand要求new_env一定要不长于old_env，不太适用

**Review by lingjun-cg — APPROVED** [review]

**Review by tanghaoth90 — APPROVED** [review]
It works for me.

**Pull request merged**